includes/classes/prod_cat_pulldown.php

```diff
@@ -0,0 +1,186 @@
+		) {
+
+			$this->exclude = $exclude;
+
+			$this->show_model = true;
+			$this->show_id = false;
+			$this->show_parent = false;
+			$this->show_full_path = false;
+
+			$this->set_selected = 0;
+			$this->values = [];
+
+			$this->keywords = $keywords;
+			$this->order_by = $order_by;
+			$this->option_filter = $filter_by_option_name;
+
+			$this->pulldown = 'product';
+			$this->attributes_join = '';
+
+			if ($attributes) {
+				$this->attributes_join = " LEFT JOIN " . TABLE_PRODUCTS_ATTRIBUTES . " pa USING (products_id)";
+			}
+
+			$this->condition = " WHERE pa.options_id =" . (int)$this->option_filter;
+
+			if ($pulldown == 'category') {
+				$this->pulldown = $pulldown;
+				$this->category_vars();
+			} else {
+				$this->product_vars();
+			}
+
+			$this->run_sql();
+
+			// default styling
+			$this->parameters = '';
+			//$this->parameters = 'required size="15" class="form-control" id="products_id"';
+		}
+
+		private function category_vars()
+		{
+			$this->sql = "SELECT DISTINCT c.categories_id as common_id, c.categories_id, cd.categories_name
+            FROM " . TABLE_CATEGORIES . " c
+            LEFT JOIN " . TABLE_CATEGORIES_DESCRIPTION . " cd ON (c.categories_id = cd.categories_id AND cd.language_id = " . (int)$_SESSION['languages_id'] . ")
+            LEFT JOIN " . TABLE_PRODUCTS_TO_CATEGORIES . " ptoc on (c.categories_id = ptoc.categories_id) "
+				. $this->attributes_join;
+
+			$this->sort = " ORDER BY categories_name";
+			$this->keyword_search_fields = [
+				'cd.categories_name',
+				'c.parent_id',
+				'cd.categories_description',
+				'c.categories_id',
+			];
+		}
+
+		private function product_vars()
+		{
+			$this->keyword_search_fields = [
+				'pd.products_name',
+				'p.products_model',
+				'pd.products_description',
+				'p.products_id',
+			];
+
+			$parm_2 = '%2$s ';
+			if (!$this->show_model) {
+				$parm_2 = '';
+			}
+
+			$this->sort = ' ORDER BY pd.products_name';
+			$this->output_string = '%1$s ' . $parm_2 . '(%3$s)';// format string with name first
+
+			if ($this->order_by == 'model') {
+				$this->sort = ' ORDER by p.products_model';
+				$this->output_string = (!empty($parm_2) ? $parm_2 . '-' : '') . ' %1$s (%3$s)'; // format string with model first
+			}
+
+			$this->sql = "SELECT p.products_id as common_id, pd.products_id, p.products_sort_order, p.products_price, p.products_model, ptc.categories_id
+                FROM " . TABLE_PRODUCTS . " p
+                LEFT JOIN " . TABLE_PRODUCTS_DESCRIPTION . " pd USING (products_id)
+                LEFT JOIN " . TABLE_PRODUCTS_TO_CATEGORIES . " ptc USING (products_id) "
+				. $this->attributes_join .
+				" WHERE pd.language_id = " . (int)$_SESSION['languages_id'];
+		}
+
+		private function run_sql()
+		{
+			global $db, $currencies;
+
+			$search = '';
+			if (!empty($this->keywords)) {
+				$search = zen_build_keyword_where_clause($this->keyword_search_fields,
+					trim(zen_db_input(zen_db_prepare_input($this->keywords))));
```

Comment:
What does this outer trim accomplish that is not already done?
````php
function zen_db_prepare_input($string, bool $trimspace = true)
{
    if (is_string($string)) {
        if ($trimspace == true) {
            return trim(stripslashes($string));
        } else {
            return stripslashes($string);
        }
    } elseif (is_array($string)) {
        foreach ($string as $key => $value) {
            $string[$key] = zen_db_prepare_input($value);
        }
        return $string;
    } else {
        return $string;
    }
}
````
This based on the requirement of $keywords being a string by the parameter definition means that the content entering `zen_db_input` will already be trimmed, then `zen_db_input` doesn't add spaces, so why trim all of that?

Reply:
:+1: